manager/state/store/volumes.go

```diff
@@ -0,0 +1,149 @@
+package store
+
+import (
+	"strings"
+
+	"github.com/docker/swarmkit/api"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+const tableVolume = "volume"
+
+func init() {
+	register(ObjectStoreConfig{
+		Table: &memdb.TableSchema{
+			Name: tableVolume,
+			Indexes: map[string]*memdb.IndexSchema{
+				indexID: {
+					Name:    indexID,
+					Unique:  true,
+					Indexer: api.VolumeIndexerByID{},
+				},
+				indexName: {
+					Name:    indexName,
+					Unique:  true,
+					Indexer: api.VolumeIndexerByName{},
+				},
+				indexCustom: {
+					Name:         indexCustom,
+					Indexer:      api.VolumeCustomIndexer{},
+					AllowMissing: true,
+				},
+				indexVolumeGroup: {
+					Name:    indexVolumeGroup,
+					Indexer: volumeIndexerByGroup{},
+				},
+				indexDriver: {
+					Name:    indexDriver,
+					Indexer: volumeIndexerByDriver{},
+				},
+			},
+		},
+		Save: func(tx ReadTx, snapshot *api.StoreSnapshot) error {
+			var err error
+			snapshot.Volumes, err = FindVolumes(tx, All)
+			return err
```

Comment:
Is it expected that you can receive an error in addition to `snapshot.Volumes` containing partial results?  The callback code in `FindVolumes` seems to hint at this (iterating over populating `volumeList` outside of the `tx`)

Reply:
frankly i haven't the slightest idea. as far as i have been concerned, the only errors that can be returned from `FindX` functions in the store are those resulting from using invalid `By` specifiers.